planner/core/expression_rewriter.go

```diff
@@ -35,20 +36,20 @@ import (
 )
 
 // EvalSubquery evaluates incorrelated subqueries once.
-var EvalSubquery func(p PhysicalPlan, is infoschema.InfoSchema, ctx sessionctx.Context) ([][]types.Datum, error)
+var EvalSubquery func(ctx context.Context, p PhysicalPlan, is infoschema.InfoSchema, sctx sessionctx.Context) ([][]types.Datum, error)
 
 // evalAstExpr evaluates ast expression directly.
-func evalAstExpr(ctx sessionctx.Context, expr ast.ExprNode) (types.Datum, error) {
+func evalAstExpr(sctx sessionctx.Context, expr ast.ExprNode) (types.Datum, error) {
 	if val, ok := expr.(*driver.ValueExpr); ok {
 		return val.Datum, nil
 	}
 	var is infoschema.InfoSchema
-	if ctx.GetSessionVars().TxnCtx.InfoSchema != nil {
-		is = ctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
+	if sctx.GetSessionVars().TxnCtx.InfoSchema != nil {
+		is = sctx.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema)
 	}
-	b := NewPlanBuilder(ctx, is)
-	fakePlan := LogicalTableDual{}.Init(ctx)
-	newExpr, _, err := b.rewrite(expr, fakePlan, nil, true)
+	b := NewPlanBuilder(sctx, is)
+	fakePlan := LogicalTableDual{}.Init(sctx)
+	newExpr, _, err := b.rewrite(context.TODO(), expr, fakePlan, nil, true)
```

Comment:
can we directly pass `ctx`?

Reply:
I prefer `context.TODO()` here because:

* `evalAstExpr` has no `context.Context` parameter and adding that involves quite a lot of changes
* `evalAstExpr` doesn't accept subquery, so it's not necessary to add the `context.Context` parameter to support subquery here.